Sources/Plasma/CoreLib/plFileSystem.cpp

```diff
@@ -58,7 +58,16 @@ You can contact Cyan Worlds, Inc. by email legal@cyan.com
 #   include <limits.h>
 #   include <sys/types.h>
 #   include <unistd.h>
+#   include <wordexp.h>
 #endif
+
+#if HS_BUILD_FOR_APPLE
+#   if __has_include(<sysdir.h>)
```

Comment:
Is there any case where the `sysdir.h` header does not exist? If so, can we depend on clang supporting `__has_include()` in all clang variants that advertise C++17 support? I'm not super familiar with clang's conformance.

Reply:
`sysdir.h` was introduced in macOS 10.12, we need to use `NSSystemDirectories.h` prior to that (which is deprecated, but still exists in current macOS, so we _could_ rely on that entirely and see if Apple ever actually removes it)

Reply:
Probably a larger conversation - I'm not sure what macOS ranges we're interested in supporting. Metal doesn't start until 10.12, so it would be the OpenGL renderer before that. The gaming input APIs don't exist until 11.0. I'd figured we might have to at runtime figure out if we can use those APIs or fallback to NSEvent. macOS 10.12 seems like a pretty reasonable minimum to me. But I'm not sure what the project goals are.

If Plasma wants to support older OSs, I am guessing NSSystemDirectories probably isn't going anywhere for a long while. It's pretty deeply used in a lot of apps.

Reply:
macOS 11 was released in late 2020, so I'm hesitant to set that as our minimum. I'm fine with using NSSystemDirectories.h if we're reasonably confident it won't disappear.

Testing the existence of sysdir.h with CMake during configuring is probably another option if we really care to do that.

Reply:
Only thing I'm checking is the status of that API on Apple Silicon. Apple sometimes uses arch transitions to cull deprecated API, but they took a pretty gentle approach on Apple Silicon so I'm betting it's still there.

Reply:
> Testing the existence of sysdir.h with CMake during configuring is probably another option if we really care to do that.

That still won't help you support < 10.12 - better use only NSSystemDirectories if you want to support those versions of macOS.

Reply:
NSSearchPathForDirectoriesInDomains is also the other common API used, and it looks like it's not deprecated. But it looks like it requires Obj-C even though it's a C function. The return types are NSString. :(
https://developer.apple.com/documentation/foundation/1414224-nssearchpathfordirectoriesindoma?language=objc

Reply:
I tried building for an earlier version of macOS, and found there might be issues there anyway. When I try to build back to 10.11 I get a bunch of errors that a lot of the modern C++ features didn't appear until macOS 10.15.

`/Users/colincornaby/Desktop/PlasmaLibs/include/string_theory/st_string.h:1193:26: 'path' is unavailable: introduced in macOS 10.15`

`/Users/colincornaby/Desktop/PlasmaLibs/include/string_theory/st_string.h:1198:37: 'u8path<const char *>' is unavailable: introduced in macOS 10.15`

And more than those.

So far all the errors look like they are in String Theory. I'm not sure how many of these functions we're actually using, but as far as I can tell the client in it's current state won't build further back than 10.15 anyway.

Reply:
> If so, can we depend on clang supporting `__has_include()` in all clang variants that advertise C++17 support? I'm not super familiar with clang's conformance.

Clang has supported `__has_include` as an extension since ±forever ([at least 3.3](https://releases.llvm.org/3.3/tools/clang/docs/LanguageExtensions.html#include-file-checking-macros), which is from 2013).

> FWIW, this `__has_include()` check is a build-time thing, so it doesn't really help you _run_ on an old macOS, I would assume, unless you build on that old macOS.

It is a build-time check, but the code below uses a [`__builtin_available`](https://clang.llvm.org/docs/LanguageExtensions.html#objective-c-available) check to decide whether to call the new or old API, depending on the runtime system version.

Regarding what should be the minimum macOS target, I would argue that at least macOS 10.14 should be supported - it is the last version to support 32-bit apps, so people are more likely to intentionally stay on that version. But I'm aware that supporting older systems isn't always easy, so if there's a good reason to require a newer version, I won't complain much (or if I do, I'll try to bring patches).

Reply:
So my current thinking is that we should set our minimum macOS compiler version to 10.14 (Xcode 9, last to support targeting 32-bit). That would, in theory, allow us to remove the `__has_include`.

I'd still like to try to use `__builtin_available` where it makes sense to see if we can support **running** on older versions.

Unfortunately, it's impossible to get CI builders that still support Xcode 9, so we aren't going to have a way to watch for breaking changes 😞

Reply:
We can set the deployment target on current Xcode to 10.14, and it should give most the necessary feedback. Two concerns I'd have:
- The string theory requirement that currently seems to point to 10.15 and higher being a requirement.
- Wider range of supported hardware. I've been trying to ignore Nvidia cards for testing. Metal makes me feel a bit better on older hardware in since OpenGL has usually a wider amount of driver variance. But even in Metal I have seen some vendor only shader issues.

Reply:
string_theory definitely supports older versions of macOS (at least as far back as 10.9)...  The trouble here is that using Xcode on newer versions of macOS to target older versions of macOS causes the ST feature detection to enable features it shouldn't (that is, compiling on macOS 10.12 will work perfectly fine for targeting macOS 10.12).  I'm not sure what the proper way to fix this is, since I'm not an experienced macOS developer...  However, you can certainly produce a working build by turning off the `std::filesystem` support via the `-DST_ENABLE_STL_FILESYSTEM=OFF` cmake flag.

Reply:
It might be possible to add an availability attribute to those methods, but I haven't tried it and it's going to be ugly code that hardcodes specific macOS versions 😞 https://clang.llvm.org/docs/AttributeReference.html#availability

We should probably take further discussion of this particular point over to an issue on string_theory.

Reply:
> It would probably be best to support only the latest Xcode toolchains, and use something like `-DST_ENABLE_STL_FILESYSTEM=OFF`. Supporting older toolchains can be iffy, and as the code gets into Apple Silicon support it may have to find symbols that aren't in older toolchains, or use compiler features that weren't in older toolchains.

I don't quite follow... isn't that usually handled with a configure-time feature check? That is, if the toolchain doesn't support the feature in question, you don't compile the new code path at all. Only if the toolchain is new enough you compile both the old and new code with a runtime check. Official builds would be done with a modern toolchain and so would conditionally support everything, but old toolchains could still compile the code, just without support for features newer than the toolchain.

Also, "the toolchain" doesn't have to be a big monolithic thing. Although Apple ships their SDKs along with specific Clang versions, it's not difficult to use older SDKs with newer Clang versions, which allows you to use modern language features even on older macOSes. Actual OS APIs like the controller input thing are the bigger problem, because you can't easily update/"retrofit" those like the compiler and libc++.

Reply:
Sorry I've been late to get back to this. It's probably not important to the pull request anyway, but I'll try and be quick. Typically on macOS you dynamically check for features where you can at runtime. So you'd always build with the latest toolchain, and then at runtime check for the presence of the symbol you rely on. It's how we'd handle multiple input methods if we wanted to support NSEvent and the new game controller framework at the same time. There wouldn't be two different builds, just a runtime check. Even though the game controller framework is only present on new OS versions, the linker is smart enough to not link it on older OS versions. For this reason, building against older SDKs is uncommon on macOS. If someone wanted to do a PPC or Intel 32 bit version, thats where older SDKs might appear.

Apple describes this process as weak linking here:
https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPFrameworks/Concepts/WeakLinking.html

This is a little different in since it's implicitly compile time, which happens. Unless there was a strong reason we'd need STL filesystem on later versions of macOS, my personal take is I'm fine with disabling the feature across the board. It simplifies things, and if the official build is going to support pre-10.15, it'll always have to be off in the official build anyway.

Reply:
Also sorry if any of the above is something you already know, and I just misinterpreted what you said. There's both compile time and run time constraints going on here, which makes talking about this in a pull request confused, at least for me.

Reply:
I think we're mostly talking about the same thing 🙂 Whether you do the runtime check using weak linking or `__builtin_available`/`@available` shouldn't make a big difference - either way you're checking if a certain API is usable at runtime, and both require that the API is supported by the SDK at compile time.

My suggestion was to have a compile-time check *in addition to* the runtime check, so that if the SDK *doesn't* support the new API, only the old code path is compiled and used. This would help when building the source locally on an old macOS - in that case you only have an old SDK, but don't care if your binaries won't use newer APIs (because your macOS doesn't have them anyway).

Getting back to where this discussion started... 🙂 The `__has_include` would only be needed to allow *compiling* on older macOSes, and supporting that would need further changes to the code. As it stands, I think the code further below won't compile if `<sysdir.h>` is not included, so it would be better to make it an unconditional include.